contents/blog/how-we-improved-feature-flags-resiliency.md

```diff
@@ -0,0 +1,145 @@
+
+This post is about how we did it, and what we learned along the way.
+
+Before we jump into how we improved things, it's worth listing out constraints that we can use to create solutions that otherwise wouldn't be possible.
+
+
+## Special Problem Constraints
+
+### PostHog Flags are deterministic
+
+Given a user identity, and a PostHog feature flag, the flag result is deterministic and sticky. So for example, if I roll out a flag to 30% of people, the person who's inside this window remains inside this window even if the rollout percentage goes up to 40, 80, or 100%.
+
+We guarantee this by computing a hash of the user id and the flag key.
+
+So, we never need a server to determine whether a flag belongs to a given rollout % or not. However, if a flag depends on person properties, we need to go to a database, check against known properties, and then determine if the flag matches or not.
+
+### Flags are evaluated multiple times in a session
+
+Flags are evaluated multiple times on a page. For example, whenever properties change our flag might change, and same for when the user identity changes. We can leverage this behaviour in our solution.
+
+### The Caching Problem
+
+It's not reasonable for us to cache the _results_ of flag evaluation for users because this blows up the size of the cache, doesn't work for new people we are seeing for the first time, and isn't flexible enough to leverage properties changing over time for users.
+
+More on this constraint in the resiliency section.
+
+
+## Making feature flags fast
+
+Since flags are deterministic, we technically don't need a server to evaluate them. This insight led to creating local evaluation of feature flags, where our SDKs download flag definitions, evaluate them locally, and only fallback to our servers when this is not possible.
+
+As we saw in our problem constraint, local evaluation can fail when we don't know properties the flag depends on. To combat this, the SDK interface allows passing in properties that you already know of. We then use these passed in properties to figure out if flag computation is possible.
+
+This optimisation is great because it cuts out all network I/O and makes evaluation CPU-bound, which reduces latency from 500ms to 10-20ms.
+
+One thing to note though is that this only works on server-side libraries. Flag definitions can have PII, like user email IDs, and require auth to download them, which means we can't expose these on the client-side libraries.
+
+As a result, our client still has not-great latency. To combat this issue, we introduced bootstrapped feature flags: You can pass in a user ID and flags to initialise a client-side PostHog SDK with. This ensures flags are instantly available, and unlocks creating cool features like redirecting on pageload based on feature flags.
```

Comment:
```suggestion
As a result, our client still has not-great latency. To combat this issue, we introduced bootstrapped feature flags: You can initialize a client-side PostHog SDK by passing in a client ID and flags. This ensures flags are instantly available, and unlocks creating cool features like redirecting on pageload based on feature flags.
```

I was thrown by the `with` at the end of the sentence, but maybe just me 😅

Reply:
I like yours better